api/src/opentrons/protocol_engine/state/geometry.py

```diff
@@ -311,3 +313,35 @@ def get_ancestor_slot_name(self, labware_id: str) -> DeckSlotName:
             )
 
         return slot_name
+
+    def ensure_location_not_occupied(
+        self, location: LabwareLocation
+    ) -> LabwareLocation:
+        """Ensure that the location does not already have equipment in it."""
+        if isinstance(location, (DeckSlotLocation, ModuleLocation)):
+            self._labware.raise_if_labware_in_location(location)
+            self._modules.raise_if_module_in_location(location)
+        return location
+
+    def get_labware_center(
+        self, labware_id: str, location: Union[DeckSlotLocation, ModuleLocation]
+    ) -> Point:
+        """Get the center point of the labware as placed on the given location.
+
+        Returns the absolute position of the labware as if it were placed on the
+        specified location. Labware offset not included.
+        """
+        labware_dimensions = self._labware.get_dimensions(labware_id)
+        module_offset = LabwareOffsetVector(x=0, y=0, z=0)
+        location_slot: DeckSlotName
+        if isinstance(location, ModuleLocation):
+            module_offset = self._modules.get_module_offset(location.moduleId)
+            location_slot = self._modules.get_location(location.moduleId).slotName
+        else:
+            location_slot = location.slotName
+        slot_center = self._labware.get_slot_center_position(location_slot)
+        return Point(
+            slot_center.x + module_offset.x,
+            slot_center.y + module_offset.y,
+            slot_center.z + module_offset.z + labware_dimensions.z / 2,
+        )
```

Comment:
No action needed, but this collaborator selector has a few warnings signs that I want to call out for future reference:

- Several calls to selectors of the same store in a row
- Non-trivial arithmetic

As mentioned during our in-person review (and I think a TODO in this file already), the `GeometryView` mixes a lot of delegation and logic, which means it's hard to test and therefor prone to sneaky bugs.

Looking at the selectors and information in play, I think I might have broken this method down a different way:

```suggestion
        # get module offset, if applicable
        module_offset_and_slot = (
            self._modules.get_slot_and_labware_offset(location.moduleId)
            if isinstance(location, ModuleLocation)
            else None
        )
        
        # get labware center, taking optional module offset and slot into account
        # internally, reads slot definition and labware definition
        # and adds things up
        return self._labware.get_labware_center(labware_id, module_offset_and_slot)
```